### ISSUE TITLE
return break try/finally block in rules

### ISSUE THREAD

**Issue by gduprey** [openhab/openhab#2339]
Using a 'return' statement within a rule breaks the contract of a try/finally block.

Consider this example:

try {
  // Get concurrency lock
  theLock.lock()
 
  // Some logic here
  
  // Use of return
  return true
  
} finally {
  // WILL NOT BE CALLED
  theLock.unlock()
}

In this case, the unlock in the finally clause will not be called.  This leaves the concurrency lock in a locked state and all subsequent attempts to run the rule will block forever.

**Comment by Seb-Maier** [comment]
It's not only a return statement, it could also be a error a function/script

**Comment by gduprey** [comment]
Yikes -- so basically, try/finally are entirely un-useful.  Their entire reason to exist is to ensure a block of code always executes to allow cleanup and such.  I would hope either this can be fixed and try/finally works as you'd reasonably expect or it was just removed from the language entirely (it's incredibly dangerous to have something so bound into code reliability not work)

**Comment by Seb-Maier** [comment]
Maybe this is a workaround

try {
// Get concurrency lock
theLock.lock()
tLock= createTimer(now.plusSeconds(iTimeout)) [|
		theLock.unlock()
	]

// Some logic here

// Use of return
return true

} finally {
// WILL NOT BE CALLED
if(tLock != null)
{
	tLock.cancel
	tLock = null
}
theLock.unlock()
}

**Comment by gduprey** [comment]
My example was just an attempt to show the problem.  While we can sort of work around it, the entire purpose of try/finally is to have a "contract" that a block of code will always run now matter how the related block concludes.  

The timer approach isn't a bad one, but there are plenty of cases where a timer would not be enough (including a case where the createTimer() call itself failed and left the lock in place).

In short, if try/finally doesn't work, you're better off without using it at all (using it may suggest to the next programmer that there is something in place where there isn't)

**Comment by teichsta** [comment]
Thanks a lot for sharing this issue and the detailed analysis @gduprey. However, you probably know that 1.8 has been the last release of the 1.x core which the rule engine belongs to. The next release (1.9) will use the ESH core. Hence i doubt it is worth spending time getting this very core changes done in 1.8..
